src/backend/loadContext.js

```diff
@@ -0,0 +1,115 @@
+  +weightedGraph: WeightedGraphT,
+  +cred: TimelineCred,
+|};
+
+export type LoadContextOptions = {|
+  +cache: CacheProvider,
+  +reporter: TaskReporter,
+  +githubToken: ?GithubToken,
+|};
+
+/**
+ * This class is responsible composing all the variables and concrete functions
+ * of the loading process.
+ *
+ * Specifically it composes:
+ * - The loading environment (through the constructor).
+ * - Concrete functions of the loading process (internally).
+ * - Parameters for a load (Project and TimelineCredParameters).
+ *
+ * You can think of LoadContext as an instance where the environment and
+ * functions have been composed so it's ready to perform a load with.
+ */
+export class LoadContext {
+  +_options: LoadContextOptions;
+
+  constructor(opts: LoadContextOptions) {
+    this._options = opts;
+  }
+
+  /**
+   * Here we're exposing multiple "proxy functions".
+   * They're just aliases of functions from another module. But by aliasing them
+   * as private properties we allow test code to spyOn these per LoadContext
+   * instance, and without needing to know details of the external modules.
+   *
+   * Of course this would break if the external module changes, but that would
+   * also occur if we directly depended on them.
+   */
+
+  +_declarations = PluginLoaders.declarations;
+  +_updateMirror = PluginLoaders.updateMirror;
+  +_createPluginGraphs = PluginLoaders.createPluginGraphs;
+  +_contractPluginGraphs = PluginLoaders.contractPluginGraphs;
+  +_overrideWeights = WeightedGraph.overrideWeights;
+  +_computeTask = ComputeFunction.computeTask;
+
+  /**
+   * The above proxy functions we're deferring to, accept interfaces so they
+   * could easily be mocked. This class takes the role of composing the concrete
+   * implementations though. So we're exposing them as aliases here, similar to
+   * the functions. As we'll need to test if these have been correctly passed on.
+   */
+
+  +_compute: ComputeFunctionT = TimelineCred.compute;
+  +_pluginLoaders: PluginLoadersT = {
+    github: githubLoader,
+    discourse: discourseLoader,
+    identity: identityLoader,
+  };
+
+  /**
+   * Performs a load in this context.
+   */
+  async load(
+    project: Project,
+    params: ?$Shape<TimelineCredParameters>,
+    weightsOverrides: WeightsT
+  ): Promise<LoadResult> {
+    const cachedProject = await this._updateMirror(
+      this._pluginLoaders,
+      this._options,
+      project
+    );
+    const pluginGraphs = await this._createPluginGraphs(
+      this._pluginLoaders,
+      this._options,
+      cachedProject
+    );
+    const contractedGraph = await this._contractPluginGraphs(
+      this._pluginLoaders,
+      pluginGraphs
+    );
+    const weightedGraph = this._overrideWeights(
+      contractedGraph,
+      weightsOverrides
+    );
+    const plugins = this._declarations(this._pluginLoaders, project);
+    const cred = await this._computeTask(this._compute, this._options, {
+      params: params || {},
```

Comment:
we shouldn't need `|| {}` -- the type signature for computTask already allows null-ish params

Reply:
I'm afraid not. This is another case where we're mismatching:
`key?: $Shape<T>` (what `.compute()` uses)
`key: ?$Shape<T>` (what `load` uses)

However. It's a good point, I want to fix that at this invocation and move this problem to the `api/load.js` file.

Reply:
Addressed by updating signature here. See https://github.com/sourcecred/sourcecred/pull/1621#issuecomment-582116351